### PR TITLE
Make string._stringLength readonly

### PR DESCRIPTION
**Pull request by benaadams**
/cc @jkotas @EgorBo

### CODE REVIEW

**Comment by stephentoub** [comment]
Is this just cleanliness or does it improve something else?

**Comment by benaadams** [comment]
Just cleanliness as string length should be immutable; it not being `readonly` gives the wrong signal.

**Comment by jkotas** [comment]
`_firstChar` is read-only too. Does `_length` marked readonly and `_firstChar` not being marked readonly send even more wrong signal?

I know marking `_firstChar` readonly would result into an extra ceremony that I am not fan of. It would be nice to add as comment at least.

**Comment by benaadams** [comment]
Well `_firstChar` is used in `fixed` statements which then mutate the string, so it isn't entirely readonly?

**Comment by jkotas** [comment]
Both are conceptually readonly in the same way. 

Some of the use in fixed statements are because string constructors have to be written in a special way and the C# compiler does not understand that the method is actually a string constructor where mutation is allowed. It is accidental that the `_length` is not assigned to in these special string constructors. (IIRC, it was assigned to at some point in the past.)

And some of the use in fixed statements is because we take shortcuts in CoreLib.

**Comment by VSadov** [comment]
`fixed` should not care about `readonly`, but it looks some of the use of `_firstChar` is ordinary `ref`

**Comment by benaadams** [comment]
Added comment

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 1]
```diff
@@ -31,8 +31,9 @@ public sealed partial class String : IComparable, IEnumerable, IConvertible, IEn
         [NonSerialized]
         private readonly int _stringLength;
 
-        // For empty strings, this will be '\0' since
-        // strings are both null-terminated and length prefixed
+        // For empty strings, this will be '\0' since strings are both null-terminated and length prefixed.
+        // Also readonly, however String uses .ctors that C# doesn't recognise as .ctors;
+        // so it complains about modifitcation outside a constructor.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/String.cs]:
```suggestion
        // so it complains about modification outside a constructor.
```

Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/String.cs]: I need to install a spell checker extension :)

**Review by jkotas — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Pull request merged**